akka-projection-core/src/main/scala/akka/projection/HandlerRecovery.scala

```diff
@@ -33,56 +22,76 @@ object HandlerRecoveryStrategy {
    * same envelope this number of `retries` with the `delay` between each attempt. It will give up
    * and fail the stream if all attempts fail.
    */
-  def retryAndFail(retries: Int, delay: FiniteDuration): HandlerRecoveryStrategy =
+  def retryAndFail(retries: Int, delay: FiniteDuration): AtLeastOrExactlyOnceStrategies =
     if (retries < 1) fail else RetryAndFail(retries, delay)
 
   /**
    * Java API: If the first attempt to invoke the handler fails it will retry invoking the handler with the
    * same envelope this number of `retries` with the `delay` between each attempt. It will give up
    * and fail the stream if all attempts fail.
    */
-  def retryAndFail(retries: Int, delay: java.time.Duration): HandlerRecoveryStrategy =
+  def retryAndFail(retries: Int, delay: java.time.Duration): AtLeastOrExactlyOnceStrategies =
     retryAndFail(retries, delay.asScala)
 
   /**
    * Scala API: If the first attempt to invoke the handler fails it will retry invoking the handler with the
    * same envelope this number of `retries` with the `delay` between each attempt. It will give up,
    * discard the element and continue with next if all attempts fail.
    */
-  def retryAndSkip(retries: Int, delay: FiniteDuration): HandlerRecoveryStrategy =
+  def retryAndSkip(retries: Int, delay: FiniteDuration): AtLeastOrExactlyOnceStrategies =
     if (retries < 1) fail else RetryAndSkip(retries, delay)
 
   /**
    * Java API: If the first attempt to invoke the handler fails it will retry invoking the handler with the
    * same envelope this number of `retries` with the `delay` between each attempt. It will give up,
    * discard the element and continue with next if all attempts fail.
    */
-  def retryAndSkip(retries: Int, delay: java.time.Duration): HandlerRecoveryStrategy =
+  def retryAndSkip(retries: Int, delay: java.time.Duration): AtLeastOrExactlyOnceStrategies =
     retryAndSkip(retries, delay.asScala)
 
   /**
    * INTERNAL API: placed here instead of the `internal` package because of sealed trait
    */
   @InternalApi private[akka] object Internal {
-    case object Fail extends HandlerRecoveryStrategy
-    case object Skip extends HandlerRecoveryStrategy
-    final case class RetryAndFail(retries: Int, delay: FiniteDuration) extends HandlerRecoveryStrategy {
+    sealed trait AtMostOnceRecoveryStrategy extends HandlerRecoveryStrategy
+    sealed trait AtLeastOnceRecoveryStrategy extends HandlerRecoveryStrategy
+    sealed trait ExactlyOnceRecoveryStrategy extends HandlerRecoveryStrategy
+
+    final type AllRecoveryStrategies = AtMostOnceRecoveryStrategy
+      with AtLeastOnceRecoveryStrategy
+      with ExactlyOnceRecoveryStrategy
+
+    final type AtLeastOrExactlyOnceStrategies = AtLeastOnceRecoveryStrategy with ExactlyOnceRecoveryStrategy
+
+    case object Fail
+        extends AtMostOnceRecoveryStrategy
+        with AtLeastOnceRecoveryStrategy
+        with ExactlyOnceRecoveryStrategy
```

Comment:
I'm not sure we need types per delivery semantics. I identify two kinds, `Strict` and `Retries`. 

`Strict` can be used everywhere, while `Retries` only for at-least-once and exactly-once. 

at-most-once can only accept `Strict`, the others accept both. 

The names may not be intuitive though.

Reply:
If we take that path, we only need: 

```scala
seated trait RecoveryStrategy
seated trait StrictRecoveryStrategy extends RecoveryStrategy
seated trait RetryRecoveryStrategy extends RecoveryStrategy {
  def etries: Int
  def delay: FiniteDuration
}

case object Fail extends StrictRecoveryStrategy
case object Skip extends StrictRecoveryStrategy

final case class RetryAndFail(retries: Int, delay: FiniteDuration) extends RecoveryStrategy
final case class RetryAndSkip(retries: Int, delay: FiniteDuration) extends RecoveryStrategy
```

and then

```scala
trait AtMostOnceCassandraProjection {
  def withRecoveryStrategy(recoveryStrategy: StrictRecoveryStrategy) 
}

trait AtLeastOnceCassandraProjection {
  def withRecoveryStrategy(recoveryStrategy: RecoveryStrategy) 
}
```

Reply:
I was thinking the same, but couldn't come up with a good name.  I thought it might be useful to have settings per message delivery type, but I think your suggestions fit best for now.